<a href="https://colab.research.google.com/github/akhiltabjula/ComputationalMethodsForInformationSystems/blob/master/In_class_exercise_09.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **The ninth in-class-exercise (20 points in total, 11/11/2020)**

The purpose of the exercise is to practice different machine learning algorithms for text classification as well as the performance evaluation. In addition, you are requried to conduct *10 fold cross validation (https://scikit-learn.org/stable/modules/cross_validation.html)* in the training. 

The dataset can be download from here: https://github.com/unt-iialab/INFO5731_FALL2020/blob/master/In_class_exercise/exercise09_datacollection.zip. The dataset contains two files train data and test data for sentiment analysis in IMDB review, it has two categories: 1 represents positive and 0 represents negative. You need to split the training data into training and validate data (80% for training and 20% for validation, https://towardsdatascience.com/train-test-split-and-cross-validation-in-python-80b61beca4b6) and perform 10 fold cross validation while training the classifier. The final trained model was final evaluated on the test data. 

Algorithms:

(1) MultinominalNB

(2) SVM 

(3) KNN 

(4) Decision tree

(5) Random Forest

(6) XGBoost

Evaluation measurement:

(1) Accuracy

(2) Recall

(3) Precison 

(4) F-1 score

In [ ]:
# Write your code here
import pandas as pd
import numpy as np
import sklearn as skl
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

In [ ]:
data_df = pd.read_csv(r'/content/stsa-train.txt',sep = 'delimiter',header= None,names = ["review"])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


In [ ]:
data_df.head(2)

review
0  1 a stirring , funny and finally transporting ...
1  0 apparently reassembled from the cutting-room...

In [11]:
data_df[['polarity','review']] = data_df["review"].str.split(" ", 1, expand=True)

In [12]:
data_df.head(2)

review polarity
0     no movement , no yuks , not much of anything .        0
1  a gob of drivel so sickly sweet , even the eag...        0

In [25]:
test_df = pd.read_csv(r'/content/stsa-test.txt',sep = 'delimiter',header= None,names = ["review"])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


In [26]:
test_df.head(2)

review
0   0 no movement , no yuks , not much of anything .
1  0 a gob of drivel so sickly sweet , even the e...

In [27]:
test_df[['polarity','review']] = test_df["review"].str.split(" ", 1, expand=True)

In [28]:
test_df.head(2)

review polarity
0     no movement , no yuks , not much of anything .        0
1  a gob of drivel so sickly sweet , even the eag...        0

In [29]:
#data cleaning
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
wn= WordNetLemmatizer()
nltk.download('wordnet')
stop_words = set(stopwords.words('english')) 
import string
import re


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


In [31]:
def data_clean(ip):
  ip = " ".join(i.lower() for i in ip.split(" ") if i not in string.punctuation)
  ip = "".join(i for i in ip if i.isalnum() or i == " ")
  values = re.split("\W+",ip)
  ip = [wn.lemmatize(i) for i in values if i not in stop_words]
  return ip

In [32]:
data_clean("hello world $")

['hello', 'world']

**Training Data**

In [36]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidfVector = TfidfVectorizer(analyzer=data_clean)
tf_idf_vector = tfidfVector.fit_transform(data_df["review"])

In [37]:
tf_idf_df = pd.DataFrame(tf_idf_vector.toarray())
tf_idf_df.columns = tfidfVector.get_feature_names()

**Testing Data**

In [40]:

tfidfVector1 = TfidfVectorizer(analyzer=data_clean,vocabulary=tfidfVector.vocabulary_)
tfidfVector1.fit(test_df["review"])
tf_idf_vector_test = tfidfVector1.transform(test_df["review"])

In [96]:
tf_idf_df_test = pd.DataFrame(tf_idf_vector_test.toarray())
tf_idf_df_test.columns = tfidfVector1.get_feature_names()

**training and testing data**

In [44]:
x_train,x_test,y_train,y_test = train_test_split(tf_idf_df, data_df["polarity"].values,test_size=0.2)

#Multinomial NB

In [45]:
#MultinomialNB

multiNB = MultinomialNB()
model_multinb = multiNB.fit(x_train,y_train)

In [46]:
#prediction for training data
multiNB_output = model_multinb.predict(x_test)

In [47]:
#prediction for testing data
multiNB_output_test = model_multinb.predict(tf_idf_vector_test)

#Result

**training validation**

In [ ]:
print("Accuracy Score :", accuracy_score(multiNB_output,y_test))

Accuracy Score : 0.7680635838150289


In [ ]:
print("Classification result")
print(classification_report(y_test,multiNB_output))

Classification result
              precision    recall  f1-score   support

           0       0.79      0.69      0.74       658
           1       0.75      0.84      0.79       726

    accuracy                           0.77      1384
   macro avg       0.77      0.76      0.77      1384
weighted avg       0.77      0.77      0.77      1384



In [ ]:
#using kfolds

print("cross validation score")
cross_val_score(model_multinb,x_test,y_test,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.6943384422896466

**testing prediction**

In [48]:
print("Accuracy Score :", accuracy_score(multiNB_output_test,test_df["polarity"]))

Accuracy Score : 0.9368478857770456


In [49]:
print("Classification result")
print(classification_report(test_df["polarity"],multiNB_output_test))

Classification result
              precision    recall  f1-score   support

           0       0.95      0.92      0.94       912
           1       0.93      0.95      0.94       909

    accuracy                           0.94      1821
   macro avg       0.94      0.94      0.94      1821
weighted avg       0.94      0.94      0.94      1821



In [53]:
#using kfolds
print("cross validation score")
cross_val_score(model_multinb,tf_idf_vector_test,test_df["polarity"].values,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.7325767129045817

#SVM

In [55]:
#Linear SVC

svm = LinearSVC()
#train the model

svm_model = svm.fit(x_train,y_train)

In [56]:
#predict the output
svm_output = svm_model.predict(x_test)

In [58]:
#predict the testing data output
svm_output_test = svm_model.predict(tf_idf_vector_test)

#Results

**training validation**

In [ ]:
print("Accuracy Score :", accuracy_score(svm_output,y_test))

Accuracy Score : 0.7622832369942196


In [ ]:
print("Classification result")
print(classification_report(y_test,svm_output))

Classification result
              precision    recall  f1-score   support

           0       0.74      0.76      0.75       658
           1       0.78      0.76      0.77       726

    accuracy                           0.76      1384
   macro avg       0.76      0.76      0.76      1384
weighted avg       0.76      0.76      0.76      1384



In [ ]:
#using kfolds
print("cross validation score")
cross_val_score(svm_model,x_test,y_test,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.7059013658638307

**testing prediction**

In [60]:
print("Accuracy Score :", accuracy_score(svm_output_test,test_df["polarity"].values))

Accuracy Score : 0.943437671609006


In [61]:
print("Classification result")
print(classification_report(test_df["polarity"].values,svm_output_test))

Classification result
              precision    recall  f1-score   support

           0       0.95      0.94      0.94       912
           1       0.94      0.95      0.94       909

    accuracy                           0.94      1821
   macro avg       0.94      0.94      0.94      1821
weighted avg       0.94      0.94      0.94      1821



In [62]:
#using kfolds
print("cross validation score")
cross_val_score(svm_model,tf_idf_vector_test,test_df["polarity"].values,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.7205008106647451

#KNN 

In [69]:
knn = KNeighborsClassifier(algorithm='brute')

In [70]:
#train the model
knn_model = knn.fit(x_train,y_train)

In [71]:
#predict the output
knn_output = knn_model.predict(x_test)

In [72]:
#predict the testing output
knn_output_test = knn_model.predict(tf_idf_vector_test)

#Result

**training validation**

In [ ]:
print("Accuracy Score :", accuracy_score(knn_output,y_test))

Accuracy Score : 0.48265895953757226


In [ ]:
print("Classification result")
print(classification_report(y_test,knn_output))

Classification result
              precision    recall  f1-score   support

           0       0.48      0.99      0.65       658
           1       0.78      0.02      0.04       726

    accuracy                           0.48      1384
   macro avg       0.63      0.51      0.34      1384
weighted avg       0.64      0.48      0.33      1384



In [ ]:
#using kfolds
print("cross validation score")
cross_val_score(knn_model,x_test,y_test,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.6712543009071005

**testing prediction**

In [73]:
print("Accuracy Score :", accuracy_score(knn_output_test,test_df["polarity"].values))

Accuracy Score : 0.6919275123558485


In [74]:
print("Classification result")
print(classification_report(test_df["polarity"].values,knn_output_test))

Classification result
              precision    recall  f1-score   support

           0       0.62      0.97      0.76       912
           1       0.94      0.41      0.57       909

    accuracy                           0.69      1821
   macro avg       0.78      0.69      0.66      1821
weighted avg       0.78      0.69      0.66      1821



In [76]:
#using kfolds
print("cross validation score")
cross_val_score(knn_model,tf_idf_vector_test,test_df["polarity"].values,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.6123070918152885

#Decision Tree

In [77]:
dtc = DecisionTreeClassifier()

In [78]:
#fit the model 
dtc_model = dtc.fit(x_train,y_train)

In [ ]:
#predict the model
dtc_output = dtc_model.predict(x_test)

In [79]:
#predict the test data
dtc_output_test = dtc_model.predict(tf_idf_vector_test)

#Results

**Training Validation**

In [ ]:
print("Accuracy Score :", accuracy_score(dtc_output,y_test))

Accuracy Score : 0.6510115606936416


In [ ]:
print("Classification result")
print(classification_report(y_test,dtc_output))

Classification result
              precision    recall  f1-score   support

           0       0.62      0.68      0.65       658
           1       0.68      0.62      0.65       726

    accuracy                           0.65      1384
   macro avg       0.65      0.65      0.65      1384
weighted avg       0.65      0.65      0.65      1384



In [ ]:
#using kfolds

print("cross validation score")
cross_val_score(dtc_model,x_test,y_test,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.6076582212490876

**Testing Validation**

In [80]:
print("Accuracy Score :", accuracy_score(dtc_output_test,test_df["polarity"].values))

Accuracy Score : 0.9302580999450851


In [81]:
print("Classification result")
print(classification_report(test_df["polarity"].values,dtc_output_test))

Classification result
              precision    recall  f1-score   support

           0       0.92      0.94      0.93       912
           1       0.94      0.92      0.93       909

    accuracy                           0.93      1821
   macro avg       0.93      0.93      0.93      1821
weighted avg       0.93      0.93      0.93      1821



In [82]:
#using kfolds

print("cross validation score")
cross_val_score(dtc_model,tf_idf_vector_test,test_df["polarity"].values,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.6320572869753198

#Random Forest

In [83]:
rf = RandomForestClassifier()

In [84]:
#train the model
rf_model = rf.fit(x_train,y_train)

In [ ]:
#predict the ouput
rf_output = rf.predict(x_test)

In [85]:
#predict the test data
rf_output_test = rf.predict(tf_idf_vector_test)

#Results

**Training validation**

In [ ]:
print("Accuracy Score :", accuracy_score(rf_output,y_test))

Accuracy Score : 0.7276011560693642


In [ ]:
print("Classification result")
print(classification_report(y_test,rf_output))

Classification result
              precision    recall  f1-score   support

           0       0.70      0.76      0.73       658
           1       0.76      0.70      0.73       726

    accuracy                           0.73      1384
   macro avg       0.73      0.73      0.73      1384
weighted avg       0.73      0.73      0.73      1384



In [ ]:
#using kfolds

print("cross validation score")
cross_val_score(rf_model,x_test,y_test,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.6640131373162339

**Testing Prediction**

In [86]:
print("Accuracy Score :", accuracy_score(rf_output_test,test_df["polarity"].values))

Accuracy Score : 0.9362987369577156


In [87]:
print("Classification result")
print(classification_report(test_df["polarity"].values,rf_output_test))

Classification result
              precision    recall  f1-score   support

           0       0.91      0.96      0.94       912
           1       0.96      0.91      0.93       909

    accuracy                           0.94      1821
   macro avg       0.94      0.94      0.94      1821
weighted avg       0.94      0.94      0.94      1821



In [88]:
#using kfolds

print("cross validation score")
cross_val_score(rf_model,tf_idf_vector_test,rf_output_test,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.6918963550111091

#XGBoost

In [89]:
xgb = XGBClassifier()

In [92]:
#train the model
xgb_model = xgb.fit(x_train,y_train)

In [91]:
#predict the model
xgb_output = xgb.predict(x_test)

In [97]:
#predict the testing data
xgb_output_test = xgb.predict(tf_idf_df_test)

#Results

**Training Validation**

In [ ]:
print("Accuracy Score :", accuracy_score(xgb_output,y_test))

Accuracy Score : 0.6495664739884393


In [ ]:
print("Classification result")
print(classification_report(y_test,xgb_output))

Classification result
              precision    recall  f1-score   support

           0       0.73      0.42      0.53       658
           1       0.62      0.86      0.72       726

    accuracy                           0.65      1384
   macro avg       0.67      0.64      0.63      1384
weighted avg       0.67      0.65      0.63      1384



In [101]:
#using kfolds
print("cross validation score")
cross_val_score(xgb_model,x_test,y_test,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.574924924924925

**Testing Prediction**

In [99]:
print("Accuracy Score :", accuracy_score(xgb_output_test,test_df["polarity"].values))

Accuracy Score : 0.7084019769357496


In [100]:
print("Classification result")
print(classification_report(test_df["polarity"].values,xgb_output_test))

Classification result
              precision    recall  f1-score   support

           0       0.80      0.56      0.66       912
           1       0.66      0.86      0.75       909

    accuracy                           0.71      1821
   macro avg       0.73      0.71      0.70      1821
weighted avg       0.73      0.71      0.70      1821



In [102]:
#using kfolds
print("cross validation score")
cross_val_score(xgb_model,tf_idf_vector_test,xgb_output_test,cv=KFold(10,random_state = 7, shuffle= True)).mean()

cross validation score


0.8577913889389299

**Summary of accuracy comparision**

1. MultinomialNB : 0.7680635838150289
2. SVM : 0.7622832369942196
3. KNN : 0.48265895953757226
4. Decision Tree : 0.6510115606936416
5. Random Forest : 0.7276011560693642
6. XGBoost : 0.6495664739884393